### PR TITLE
Fix double apply on update default values on volatile functions

### LINKED ISSUE #56022: apply-on-update-default-values applied (again) on pressing OK
### What is the bug or the crash?

On pressing `OK` the values (that are already evaluated on editing the form) are reevaluated again.

#### Summary
When identifying an existing feature and change something, the default values are reevaluated with the setting "apply on update" for fields with expressions containing:

- a field that is currently changed (dependency/referencedColumn)
- volatile functions like `now()`, `rand()`, `randf()` or `uuid()`

That's good and fixed (since 3.34.4) with #https://github.com/qgis/QGIS/pull/55633 

But (**the issue**) on pressing `OK` those values (and this especially concerning the volatile functions) are reevaluated again.

Thoughts: Can we just turn it off to update on `OK`, or could it be, that a value should update that has not updated before by being a volatile function or having a dependency)? Or are those values not updated anyway?

### Steps to reproduce the issue

- Create a new feature in the test project
- enter a value and see that the other values (since 3.34.4. the `uuid` and the `now` and with "all" versions the field `rand`) are updated
- identify the feature
- change the value in field `value` and see that the other values (since 3.34.4. the `uuid` and the `now` and with "all" versions the field `rand`) are updated
- press OK -> and see that the `rand` value is wrongly updated again!

### Versions

**master** on 908dc2a734c65425586d533b5e67e2002119187f

QGIS version | 3.28.10-Firenze | QGIS code revision | e2edb9c468b
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.10.12
GDAL/OGR version | 3.4.1
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
GEOS version | 3.10.2-CAPI-1.16.0
SQLite version | 3.37.2
PDAL version | 2.3.0
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Ubuntu 22.04.2 LTS
  |   |   |  
Active Python plugins
ordered_relation_editor | 0.2.7
linking_relation_editor | v1.1.9
document_management_system | 0.3.2
QgisModelBaker | v7.8.1
solocator | 1.6.0
grassprovider | 2.12.99
sagaprovider | 2.12.99
processing | 2.12.99
MetaSearch | 0.3.6
db_manager | 0.1.20

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

[testproject.zip](https://github.com/qgis/QGIS/files/14061847/testproject.zip)

### PR DESCRIPTION
**Pull request by signedav**
Don't update the widgets on volatile functions in real time. Only update those fields on saving the form (pressig ok) when it's on `*EditMode` (on `AddFeatureMode` it does not update it on save anyway because they are set on form open (like normal default values)).

This fixes #56022 and reverts a part of this fix here #55633 and brings back a behavior that used to be before this bug #54446. It's what originally :

- In attribute table: Default Values are emitted but we don't see it in the widget (but they are updated closing form)
- In form via identify: Default Values are emitted but we don't see it in the widget (but they are updated closing form)

(maybe this confusing concept could be improved by having a marker / icon there, but not as part of this PR here).

### CODE REVIEW

File: tests/src/python/test_qgsattributeform.py [hunk 1]
```diff
@@ -163,7 +163,7 @@ def test_on_update(self):
 
         layer = QgsVectorLayer("Point?field=age:int", "vl", "memory")
         # set default value for numbers to [1, {age}], it will depend on the field age and should update
-        field = QgsField('numbers', QVariant.List, 'array')
+        field = QgsField('numbers', QVariant.List, subType=QVariant.Int)
```
Comment (nirvn) [inline tests/src/python/test_qgsattributeform.py:166]: @signedav , is that change necessary? If so, can you explain? If not, let's revert.

**Review by nirvn — COMMENTED** [review]
Look good, only one question.

We'll need to backport this to 3.34 as the original fix had been backported. Additional tests in there most welcome.

File: tests/src/python/test_qgsattributeform.py (hunk 1 above)
Comment (signedav) [reply tests/src/python/test_qgsattributeform.py:166]: Before on form.save the values are gone. In this test we dont do it so it doesnt matter, but in the next one we do. But if someone would extend this function they might run into the same problem when this subtype is still array.

Comment (signedav) [reply tests/src/python/test_qgsattributeform.py]: Okay, I change it back to avoid confusion because this PR did not change anything in behavior of that test.

**Review by nirvn — APPROVED** [review]
